src/main/java/dev/jlibra/admissioncontrol/AdmissionControl.java

```diff
@@ -0,0 +1,149 @@
+                .build();
+
+        RawTransaction rawTransaction = RawTransaction.newBuilder()
+                .setProgram(program)
+                .setExpirationTime(transaction.getExpirationTime())
+                .setGasUnitPrice(transaction.getGasUnitPrice())
+                .setMaxGasAmount(transaction.getMaxGasAmount())
+                .setSenderAccount(ByteString.copyFrom(KeyUtils.toByteArrayLibraAddress(publicKey.getEncoded())))
+                .setSequenceNumber(transaction.getSequenceNumber())
+                .build();
+
+        SignedTransaction signedTransaction = SignedTransaction.newBuilder()
+                .setRawTxnBytes(rawTransaction.toByteString())
+                .setSenderPublicKey(ByteString.copyFrom(KeyUtils.stripPublicKeyPrefix(publicKey.getEncoded())))
+                .setSenderSignature(ByteString.copyFrom(LibraHelper.signTransaction(rawTransaction, privateKey)))
+                .build();
+
+        SubmitTransactionRequest submitTransactionRequest = SubmitTransactionRequest.newBuilder()
+                .setSignedTxn(signedTransaction)
+                .build();
+
+        SubmitTransactionResponse response = stub.submitTransaction(submitTransactionRequest);
+
+        channel.shutdown();
+
+        return new SubmitTransactionResult(response.getAcStatus(), response.getMempoolStatus(), response.getVmStatus());
+    }
+
+    private ByteString readCodeFromStream(Transaction transaction) {
+        try {
+            return ByteString.readFrom(transaction.getProgram().getCode());
+        } catch (IOException e) {
+            throw new RuntimeException("Could not read move code from input stream", e);
+        }
+    }
+
+    public UpdateToLatestLedgerResult updateToLatestLedger(List<GetAccountState> arguments) {
+        ManagedChannel channel = ManagedChannelBuilder.forAddress(host, port)
+                .usePlaintext()
+                .build();
+
+        AdmissionControlBlockingStub stub = AdmissionControlGrpc.newBlockingStub(channel);
+
+        List<RequestItem> requestItems = arguments.stream().map(argument -> {
+            GetAccountStateRequest getAccountStateRequest = GetAccountStateRequest.newBuilder()
+                    .setAddress(ByteString.copyFrom(argument.getAddress()))
+                    .build();
+
+            RequestItem requestItem = RequestItem.newBuilder()
+                    .setGetAccountStateRequest(getAccountStateRequest)
+                    .build();
+            return requestItem;
+        }).collect(toList());
+
+        UpdateToLatestLedgerResponse response = stub.updateToLatestLedger(UpdateToLatestLedgerRequest.newBuilder()
+                .addAllRequestedItems(requestItems)
+                .build());
+
+        List<AccountState> accountStates = new ArrayList<>();
+
+        response.getResponseItemsList().forEach(responseItem -> {
+            accountStates.addAll(LibraHelper.readAccountStates(responseItem.getGetAccountStateResponse()
```

Comment:
Preferably this should not be side-effect with `addAll` inside `forEach`. Can you rewrite it to `.collect(toList())` instead?

Reply:
This is also related to the other branch I'm working on and this is a little tricky because when I'm sending query to the admission control, I can request several different things in one request. The class structure supports, for example, requesting an account state and transaction by sequence number in the same request item. And the return values are in a similar structure, so it might be that the same response item has a return value for another query as well. You'll probably see the problem if you look at the transaction example branch.
But I will need to investigate how the api works with several + different queries in the same request to find the best way to handle the response.